src/app/pcasts/dao/listening_histories_dao.py

```diff
@@ -3,34 +3,42 @@
 
 def create_or_update_listening_histories(episode_update_info_map, user):
   episode_ids = episode_update_info_map.keys()
-  episodes = episodes_dao.get_episodes(episode_ids, user.id)
+
   listening_histories = \
     ListeningHistory.query.filter(ListeningHistory.episode_id.in_(episode_ids),
                                   ListeningHistory.user_id == user.id).all()
   listening_histories_by_episode = \
     {lh.episode_id: lh for lh in listening_histories}
   result = []
-  for e in episodes:
+  for episode_id in episode_ids:
     try:
-      current_progress = episode_update_info_map[e.id]['current_progress']
-      percentage_listened = episode_update_info_map[e.id]['percentage_listened']
+      current_progress = \
+        episode_update_info_map[episode_id]['current_progress']
+      percentage_listened = \
+        episode_update_info_map[episode_id]['percentage_listened']
     except KeyError:
-      raise Exception('Invalid episode_id {} provided'.format(e.id))
-    real_duration = \
-      None if 'real_duration' not in episode_update_info_map[e.id] \
-      else episode_update_info_map[e.id]['real_duration']
-    if e.id in listening_histories_by_episode:
-      lh = listening_histories_by_episode[e.id]
+      raise Exception('Invalid episode_id {} provided'.format(episode_id))
+    if episode_id in listening_histories_by_episode:
+      lh = listening_histories_by_episode[episode_id]
       lh.current_progress = current_progress
       lh.percentage_listened = lh.percentage_listened + percentage_listened
     else:
       lh = \
-        ListeningHistory(episode_id=e.id, user_id=user.id,
-                         percentage_listened=percentage_listened, current_progress=current_progress)
+        ListeningHistory(episode_id=episode_id, user_id=user.id,
+                         percentage_listened=percentage_listened,
+                         current_progress=current_progress)
       db.session.add(lh)
-    if not e.real_duration and real_duration:
-      e.real_duration = real_duration
     result.append(lh)
+  # Update episode durations if needed
+  episode_ids_to_query = []
+  for episode_id, values in episode_update_info_map.iteritems():
+    if 'real_duration' in values:
+      episode_ids_to_query.append(episode_id)
+  episodes = episodes_dao.get_episodes(episode_ids_to_query, user.id)
+  for episode in episodes:
+    if not episode.real_duration_written:
+      episode.duration = episode_update_info_map[episode.id]['real_duration']
+      episode.real_duration_written = True
```

Comment:
Why is this necessary? You are sequentially reading through list. How would you go through this list twice?

Reply:
the real_duration_written might have been set to True from in previous batch